<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Kapitel 6 kNN | DataScience1</title>
<meta name="author" content="Sebastian Sauer">
<meta name="description" content="6.1 Lernsteuerung   6.1.1 Lernziele Sie sind in der Lage, einfache Klassifikationsmodelle zu spezifizieren mit tidymodels Sie kÃ¶nnen den knn-Algorithmus erlÃ¤utern Sie kÃ¶nnen den knn-Algorithmus in...">
<meta name="generator" content="bookdown 0.26.2 with bs4_book()">
<meta property="og:title" content="Kapitel 6 kNN | DataScience1">
<meta property="og:type" content="book">
<meta property="og:description" content="6.1 Lernsteuerung   6.1.1 Lernziele Sie sind in der Lage, einfache Klassifikationsmodelle zu spezifizieren mit tidymodels Sie kÃ¶nnen den knn-Algorithmus erlÃ¤utern Sie kÃ¶nnen den knn-Algorithmus in...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Kapitel 6 kNN | DataScience1">
<meta name="twitter:description" content="6.1 Lernsteuerung   6.1.1 Lernziele Sie sind in der Lage, einfache Klassifikationsmodelle zu spezifizieren mit tidymodels Sie kÃ¶nnen den knn-Algorithmus erlÃ¤utern Sie kÃ¶nnen den knn-Algorithmus in...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script><script src="libs/viz-1.8.2/viz.js"></script><link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script><script src="libs/es6shim-0.35.6/es6shim.js"></script><script src="libs/es7shim-6.0.0/es7shim.js"></script><script src="libs/graphre-0.1.3/graphre.js"></script><script src="libs/nomnoml-1.3.1/nomnoml.js"></script><script src="libs/nomnoml-binding-0.2.3/nomnoml.js"></script><script src="libs/d3-3.3.8/d3.min.js"></script><script src="libs/dagre-0.4.0/dagre-d3.min.js"></script><link href="libs/mermaid-0.3.0/dist/mermaid.css" rel="stylesheet">
<script src="libs/mermaid-0.3.0/dist/mermaid.slim.min.js"></script><script src="libs/chromatography-0.1/chromatography.js"></script><script src="libs/DiagrammeR-binding-1.0.6.1/DiagrammeR.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style-bs4.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="Grundlagen der Prognosemodellierung ğŸ”®ğŸ§°">DataScience1</a>:
        <small class="text-muted">Grundlagen der Prognosemodellierung ğŸ”®ğŸ§°</small>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">Zu diesem Buch</a></li>
<li><a class="" href="hinweise.html"><span class="header-section-number">1</span> Hinweise</a></li>
<li><a class="" href="pr%C3%BCfung.html"><span class="header-section-number">2</span> PrÃ¼fung</a></li>
<li class="book-part">Themen</li>
<li><a class="" href="statistisches-lernen.html"><span class="header-section-number">3</span> Statistisches Lernen</a></li>
<li><a class="" href="r-zweiter-blick.html"><span class="header-section-number">4</span> R, zweiter Blick</a></li>
<li><a class="" href="tidymodels.html"><span class="header-section-number">5</span> tidymodels</a></li>
<li><a class="active" href="knn.html"><span class="header-section-number">6</span> kNN</a></li>
<li><a class="" href="resampling-und-tuning.html"><span class="header-section-number">7</span> Resampling und Tuning</a></li>
<li><a class="" href="logistische-regression.html"><span class="header-section-number">8</span> Logistische Regression</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/sebastiansauer/datascience1">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="knn" class="section level1" number="6">
<h1>
<span class="header-section-number">Kapitel 6</span> kNN<a class="anchor" aria-label="anchor" href="#knn"><i class="fas fa-link"></i></a>
</h1>
<div id="lernsteuerung-3" class="section level2" number="6.1">
<h2>
<span class="header-section-number">6.1</span> Lernsteuerung<a class="anchor" aria-label="anchor" href="#lernsteuerung-3"><i class="fas fa-link"></i></a>
</h2>
<!-- Chapter Start sections: Lernziele, Literatur, Hinweise, ... -->
<div id="lernziele-4" class="section level3" number="6.1.1">
<h3>
<span class="header-section-number">6.1.1</span> Lernziele<a class="anchor" aria-label="anchor" href="#lernziele-4"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>Sie sind in der Lage, einfache Klassifikationsmodelle zu spezifizieren mit tidymodels</li>
<li>Sie kÃ¶nnen den knn-Algorithmus erlÃ¤utern</li>
<li>Sie kÃ¶nnen den knn-Algorithmus in tidymodels anwenden</li>
<li>Sie kÃ¶nnen die GÃ¼temetriken von Klassifikationsmodellen einschÃ¤tzen</li>
</ul>
</div>
<div id="literatur-4" class="section level3" number="6.1.2">
<h3>
<span class="header-section-number">6.1.2</span> Literatur<a class="anchor" aria-label="anchor" href="#literatur-4"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>Rhys, Kap. 3</li>
<li><a href="https://datasciencebook.ca/classification.html#classification">Timbers et al., Kap. 5</a></li>
</ul>
</div>
</div>
<div id="Ã¼berblick-1" class="section level2" number="6.2">
<h2>
<span class="header-section-number">6.2</span> Ãœberblick<a class="anchor" aria-label="anchor" href="#%C3%BCberblick-1"><i class="fas fa-link"></i></a>
</h2>
<p>In diesem Kapitel geht es um das Verfahren <em>KNN</em>, <em>K-NÃ¤chste-Nachbarn</em>.</p>
<p>BenÃ¶tigte R-Pakete fÃ¼r dieses Kapitel:</p>
<div class="sourceCode" id="cb204"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidymodels.tidymodels.org">tidymodels</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="intuitive-erklÃ¤rung" class="section level2" number="6.3">
<h2>
<span class="header-section-number">6.3</span> Intuitive ErklÃ¤rung<a class="anchor" aria-label="anchor" href="#intuitive-erkl%C3%A4rung"><i class="fas fa-link"></i></a>
</h2>
<p><em>K-NÃ¤chste-Nachbarn</em> (k nearest neighbors, kNN) ist ein einfacher Algorithmus des maschinellen Lernens,
der sowohl fÃ¼r Klassifikation als auch fÃ¼r numerische Vorhersage (Regression) genutzt werden kann.
Wir werden kNN als Beispiel fÃ¼r eine Klassifikation betrachten.</p>
<p>Betrachen wir ein einfÃ¼hrendes Beispiel von <span class="citation">Rhys (<a href="references.html#ref-rhys" role="doc-biblioref">2020</a>)</span>, fÃ¼r das es eine <a href="https://livebook.manning.com/book/machine-learning-for-mortals-mere-and-otherwise/chapter-3/22">Online-Quelle</a> gibt.
Stellen Sie sich vor, wir laufen durch englische Landschaft,
vielleicht die Grafschaft Kent, und sehen ein kleines Tier durch das Gras huschen.
Eine Schlange?!
In England gibt es (laut <span class="citation">Rhys (<a href="references.html#ref-rhys" role="doc-biblioref">2020</a>)</span>) nur eine giftige Schlange,
die Otter (Adder).
Eine andere Schlange, die <em>Grass Snake</em> ist nicht giftig,
und dann kommt noch der <em>Slow Worm</em> in Frage,
der gar nicht zur Familie der Schlangen gehÃ¶rt.
PrimÃ¤r interessiert uns die Frage, haben wir jetzt eine Otter gesehen?
Oder was fÃ¼r ein Tier war es?</p>
<p>Zum GlÃ¼ck wissen wir einiges Ã¼ber Schlangen bzw. schlangenÃ¤hnliche Tiere Englands.
NÃ¤mlich kÃ¶nnen wir die betreffenden Tierarten in GrÃ¶ÃŸe und AggressivitÃ¤t einschÃ¤tzen,
das ist in Abbildung <a href="knn.html#fig:slang">6.1</a> dargestellt.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:slang"></span>
<img src="img/rhys-fig3-2.jpeg" alt="Haben wir gerade eine Otter gesehen?" width="70%"><p class="caption">
Figure 6.1: Haben wir gerade eine Otter gesehen?
</p>
</div>
<p>Der Algorithmus von kNN sieht einfach gesagt vor,
dass wir schauen, welcher Tierarten Tiere mit Ã¤hnlicher AggressivitÃ¤t und GrÃ¶ÃŸe angehÃ¶ren.
Die Tierart die bei diesen â€œNachbarnâ€ hinsichtlich Ã„hnlichkeit relevanter Merkmale am hÃ¤ufigsten vertreten ist, ordnen wir die bisher unklassifizierte Beobachtung zu.</p>
<p>Etwas zugespitzt:</p>
<blockquote>
<p>Wenn es quakt wie eine Ente, lÃ¤uft wie eine Ente und aussieht wie eine Ente, dann ist es eine Ente.</p>
</blockquote>
<p>Die Anzahl <span class="math inline">\(k\)</span> der nÃ¤chsten Nachbarn kÃ¶nnen wir frei wÃ¤hlen;
der Wert wird <em>nicht</em> vom Algorithmuss bestimmt.
Solche vom Nutzi zu bestimmenden GrÃ¶ÃŸen nennt man auch <em>Tuningparameter</em>.</p>
</div>
<div id="krebsdiagnostik" class="section level2" number="6.4">
<h2>
<span class="header-section-number">6.4</span> Krebsdiagnostik<a class="anchor" aria-label="anchor" href="#krebsdiagnostik"><i class="fas fa-link"></i></a>
</h2>
<p>Betrachten wir ein Beispiel von <span class="citation">Timbers, Campbell, and Lee (<a href="references.html#ref-timbers_data_2022" role="doc-biblioref">2022</a>)</span>,
das <a href="https://datasciencebook.ca/classification.html#classification-with-k-nearest-neighbors">hier</a> frei eingesehen werden kann.</p>
<p>Die Daten sind so zu beziehen:</p>
<div class="sourceCode" id="cb205"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_url</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/UBC-DSCI/introduction-to-datascience/master/data/wdbc.csv"</span>
<span class="va">cancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">data_url</span><span class="op">)</span></code></pre></div>
<p>In diesem Beispiel versuchen wir Tumore der Brust zu klassifizieren,
ob sie einen schweren Verlauf (maligne, engl. malignant) oder einen weniger schweren Verlauf (benigne, engl. benign) erwarten lassen.
Der Datensatz ist <a href="https://datasciencebook.ca/classification.html#describing-the-variables-in-the-cancer-data-set">hier</a> nÃ¤her erlÃ¤utert.</p>
<p>Wie in Abb. <a href="knn.html#fig:cancer1">6.2</a> ersichtlich,
steht eine Tumordiagnose (malignant vs.Â benign) in AbhÃ¤ngigkeit
von Umfang (engl. perimeter) und <a href="https://de.wikipedia.org/wiki/Konvexe_und_konkave_Funktionen">KonkavitÃ¤t</a>,
die â€œGekrÃ¼mmtheit nach innenâ€.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:cancer1"></span>
<img src="chunk-img/tidymodelscancer1-1.png" alt="Streudiagramm zur EinschÃ¤tzung von Tumordiagnosen" width="70%"><p class="caption">
Figure 6.2: Streudiagramm zur EinschÃ¤tzung von Tumordiagnosen
</p>
</div>
<p>Wichtig ist, dass die Merkmale standardisiert sind, also eine identische Skalierung aufweisen,
da sonst das Merkmal mit kleinerer Skala weniger
in die Berechnung der NÃ¤he (bzw. Abstand) eingeht.</p>
<p>FÃ¼r einen neuen, bisher unklassifizierten Fall suchen nur nun nach einer Diagnose,
also nach der am besten passenden Diagnose (maligne oder benigne),
s. Abb. <a href="knn.html#fig:cancer2">6.3</a>, wieder aus <span class="citation">Timbers, Campbell, and Lee (<a href="references.html#ref-timbers_data_2022" role="doc-biblioref">2022</a>)</span>.
Ihr Quellcode fÃ¼r dieses Diagramm (und das ganze Kapitel) findet sich <a href="https://github.com/UBC-DSCI/introduction-to-datascience/blob/master/classification1.Rmd">hier</a>.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:cancer2"></span>
<img src="chunk-img/tidymodelscancer2-1.png" alt="Ein neuer Fall, bisher unklassifiziert" width="70%"><p class="caption">
Figure 6.3: Ein neuer Fall, bisher unklassifiziert
</p>
</div>
<p>Wir kÃ¶nnen zunÃ¤chst den (im euklidischen Koordinatensystem) nÃ¤chst gelegenen Fall (der â€œnÃ¤chste Nachbarâ€) betrachten,
und vereinbaren,
dass wir dessen Klasse als SchÃ¤tzwert fÃ¼r den unklassiffizierten Fall Ã¼bernehmen,
s. Abb. <a href="knn.html#fig:cancer3">6.4</a>.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:cancer3"></span>
<img src="chunk-img/tidymodelscancer3-1.png" alt="Ein nÃ¤chster Nachbar" width="70%"><p class="caption">
Figure 6.4: Ein nÃ¤chster Nachbar
</p>
</div>
<p>Betrachten wir einen anderen zu klassifizierenden Fall, s. Abb <a href="knn.html#fig:cancer4">6.5</a>.
Ob hier die Klassifikation von â€œbenignâ€ korrekt ist?
WomÃ¶glich nicht, denn viele andere Nachbarn,
die etwas weiter weg gelegen sind, gehÃ¶ren zur anderen Diagnose, malign.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:cancer4"></span>
<img src="https://datasciencebook.ca/_main_files/figure-html/05-knn-4-1.png" alt="TrÃ¼gt der nÃ¤chste Nachbar?" width="70%"><p class="caption">
Figure 6.5: TrÃ¼gt der nÃ¤chste Nachbar?
</p>
</div>
<p>Um die Vorhersage zu verbessern,
kÃ¶nnen wir nicht nur den nÃ¤chst gelegenen Nachbarn betrachten,
sondern die <span class="math inline">\(k\)</span> nÃ¤chst gelegenen, z.B. <span class="math inline">\(k=3\)</span>, s. Abb <a href="knn.html#fig:cancer5">6.6</a>.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:cancer5"></span>
<img src="https://datasciencebook.ca/_main_files/figure-html/05-knn-5-1.png" alt="kNN mit k=3" width="70%"><p class="caption">
Figure 6.6: kNN mit k=3
</p>
</div>
<p>Die Entscheidungsregel ist dann einfach eine Mehrheitsentscheidung:
Wir klassifizieren den neuen Fall entsprechend der Mehrheit in den <span class="math inline">\(k\)</span> nÃ¤chst gelegenen Nachbarn.</p>
</div>
<div id="berechnung-der-nÃ¤he" class="section level2" number="6.5">
<h2>
<span class="header-section-number">6.5</span> Berechnung der NÃ¤he<a class="anchor" aria-label="anchor" href="#berechnung-der-n%C3%A4he"><i class="fas fa-link"></i></a>
</h2>
<p>Es gibt verschiedenen Algorithmen,
um die NÃ¤he bzw. Distanz der Nachbarn zum zu klassifizieren Fall zu berechnen.</p>
<p>Eine gebrÃ¤uchliche Methode ist der <em>euklidische</em> Abstand,
der mit Pythagoras berechnet werden kann, s. Abb. <a href="knn.html#fig:pyth1">6.7</a> aus <span class="citation">Sauer (<a href="references.html#ref-modar" role="doc-biblioref">2019</a>)</span>.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:pyth1"></span>
<img src="img/distanz_crop.png" alt="Euklidischer Abstand wird mit der Regel von Pythagoras berechnet" width="70%"><p class="caption">
Figure 6.7: Euklidischer Abstand wird mit der Regel von Pythagoras berechnet
</p>
</div>
<p>Wie war das noch mal?</p>
<p><span class="math display">\[c^2 = a^2 + b^2\]</span></p>
<p>Im Beispiel oben also:</p>
<p><span class="math inline">\(c^2 = 3^2 + 4^2 = 5^2\)</span></p>
<p>Damit gilt: <span class="math inline">\(c = \sqrt{c^2} = \sqrt{5^2}=5\)</span>.</p>
<p>Im 2D-Raum ist das so einfach, dass man das (fat) mit bloÃŸem Augenschein entscheiden kann.
In mehr als 2 Dimensionen wird es aber schwierig fÃ¼r das Auge, wie ein <a href="https://datasciencebook.ca/classification.html#more-than-two-explanatory-variables">Beispiel</a> aus <span class="citation">Timbers, Campbell, and Lee (<a href="references.html#ref-timbers_data_2022" role="doc-biblioref">2022</a>)</span> zeigt.</p>
<p>Allerdings kann man den guten alten Pythagoras auch auf Dreiecke mit mehr als zwei Dimensionen anwenden, s. Abb. <a href="knn.html#fig:pyth2">6.8</a> aus <span class="citation">Sauer (<a href="references.html#ref-modar" role="doc-biblioref">2019</a>)</span>, Kap. 21.1.2.</p>
<!-- ```{r pyth2, echo = FALSE, fig.cap = "Pythagoras aufgebohrt"} -->
<!-- knitr::include_graphics("img/pythagoras_crop.png") -->
<!-- ``` -->
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:pyth2"></span>
<img src="img/pythagoras-crop.png" alt="Pythagoras in der Ebene (links) und in 3D (rechts)" width="45%"><img src="img/pythagoras2-crop.png" alt="Pythagoras in der Ebene (links) und in 3D (rechts)" width="45%"><p class="caption">
Figure 6.8: Pythagoras in der Ebene (links) und in 3D (rechts)
</p>
</div>
<p>Bleiben wir beim Beispiel von Anna und Berta und nehmen wir eine dritte Variable
hinzu (Statistikliebe).
Sagen wir, der Unterschied in dieser dritten Variable zwischen Anna und Berta betrage 2.</p>
<p>Es gilt:</p>
<p><span class="math display">\[
\begin{aligned}
e^2 &amp;= c^2 + d^2 \\
e^2 &amp;= 5^2 + 2^2 \\
e^2 &amp;= 25 + 4\\
e &amp;= \sqrt{29} \approx 5.4
\end{aligned}
\]</span></p>
</div>
<div id="knn-mit-tidymodels" class="section level2" number="6.6">
<h2>
<span class="header-section-number">6.6</span> kNN mit Tidymodels<a class="anchor" aria-label="anchor" href="#knn-mit-tidymodels"><i class="fas fa-link"></i></a>
</h2>
<div id="analog-zu-timbers-et-al." class="section level3" number="6.6.1">
<h3>
<span class="header-section-number">6.6.1</span> Analog zu Timbers et al.<a class="anchor" aria-label="anchor" href="#analog-zu-timbers-et-al."><i class="fas fa-link"></i></a>
</h3>
<p>Eine Anwendung von kNN mit Tidymodels ist in <span class="citation">Timbers, Campbell, and Lee (<a href="references.html#ref-timbers_data_2022" role="doc-biblioref">2022</a>)</span>, Kap. 5.6, <a href="https://datasciencebook.ca/classification.html#k-nearest-neighbors-with-tidymodels">hier</a> beschrieben.</p>
<p>Die Daten aus <span class="citation">Timbers, Campbell, and Lee (<a href="references.html#ref-timbers_data_2022" role="doc-biblioref">2022</a>)</span> finden sich <a href="https://github.com/UBC-DSCI/introduction-to-datascience/tree/master/data">in diesem Github-Repo</a>-</p>
<p>Die (z-transformierten) Daten zur Tumorklassifikation kÃ¶nnen <a href="https://raw.githubusercontent.com/UBC-DSCI/data-science-a-first-intro-worksheets/main/worksheet_classification1/data/clean-wdbc-data.csv">hier</a> bezogen werden.</p>
<div class="sourceCode" id="cb206"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_url</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/UBC-DSCI/introduction-to-datascience/master/data/wdbc.csv"</span>
<span class="va">cancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">data_url</span><span class="op">)</span></code></pre></div>
<p><span class="citation">Timbers, Campbell, and Lee (<a href="references.html#ref-timbers_data_2022" role="doc-biblioref">2022</a>)</span> verwenden in Kap. 5 auch noch nicht standardisierte Daten, <code>unscales_wdbc.csv</code>, die <a href="https://raw.githubusercontent.com/UBC-DSCI/introduction-to-datascience/master/data/unscaled_wdbc.csv">hier</a> als CSV-Datei heruntergeladen werden kÃ¶nnen.</p>
<div class="sourceCode" id="cb207"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cancer_unscales_path</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/UBC-DSCI/introduction-to-datascience/master/data/unscaled_wdbc.csv"</span>

<span class="va">unscaled_cancer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">cancer_unscales_path</span><span class="op">)</span> |&gt;
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Class <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/as_factor.html">as_factor</a></span><span class="op">(</span><span class="va">Class</span><span class="op">)</span><span class="op">)</span> |&gt;
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Class</span>, <span class="va">Area</span>, <span class="va">Smoothness</span><span class="op">)</span>
<span class="va">unscaled_cancer</span></code></pre></div>
<pre><code>## # A tibble: 569 Ã— 3
##    Class  Area Smoothness
##    &lt;fct&gt; &lt;dbl&gt;      &lt;dbl&gt;
##  1 M     1001      0.118 
##  2 M     1326      0.0847
##  3 M     1203      0.110 
##  4 M      386.     0.142 
##  5 M     1297      0.100 
##  6 M      477.     0.128 
##  7 M     1040      0.0946
##  8 M      578.     0.119 
##  9 M      520.     0.127 
## 10 M      476.     0.119 
## # â€¦ with 559 more rows</code></pre>
</div>
<div id="rezept-definieren" class="section level3" number="6.6.2">
<h3>
<span class="header-section-number">6.6.2</span> Rezept definieren<a class="anchor" aria-label="anchor" href="#rezept-definieren"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb209"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">uc_recipe</span> <span class="op">&lt;-</span> <span class="fu">recipe</span><span class="op">(</span><span class="va">Class</span> <span class="op">~</span> <span class="va">.</span>, data <span class="op">=</span> <span class="va">unscaled_cancer</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">uc_recipe</span><span class="op">)</span></code></pre></div>
<pre><code>## Recipe
## 
## Inputs:
## 
##       role #variables
##    outcome          1
##  predictor          2</code></pre>
<p>Und jetzt die z-Transformation:</p>
<div class="sourceCode" id="cb211"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">uc_recipe</span> <span class="op">&lt;-</span> 
  <span class="va">uc_recipe</span> |&gt;
  <span class="fu">step_scale</span><span class="op">(</span><span class="fu">all_predictors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> |&gt;
  <span class="fu">step_center</span><span class="op">(</span><span class="fu">all_predictors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Die Schritte <code>prep()</code> und <code>bake()</code> sparen wir uns, da <code>fit()</code> und <code><a href="https://rdrr.io/r/stats/predict.html">predict()</a></code>
das fÃ¼r uns besorgen.</p>
</div>
<div id="modell-definieren" class="section level3" number="6.6.3">
<h3>
<span class="header-section-number">6.6.3</span> Modell definieren<a class="anchor" aria-label="anchor" href="#modell-definieren"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb212"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">knn_spec</span> <span class="op">&lt;-</span> <span class="fu">nearest_neighbor</span><span class="op">(</span>weight_func <span class="op">=</span> <span class="st">"rectangular"</span>, neighbors <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> |&gt;
  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"kknn"</span><span class="op">)</span> |&gt;
  <span class="fu">set_mode</span><span class="op">(</span><span class="st">"classification"</span><span class="op">)</span>
<span class="va">knn_spec</span></code></pre></div>
<pre><code>## K-Nearest Neighbor Model Specification (classification)
## 
## Main Arguments:
##   neighbors = 5
##   weight_func = rectangular
## 
## Computational engine: kknn</code></pre>
</div>
<div id="workflow-definieren" class="section level3" number="6.6.4">
<h3>
<span class="header-section-number">6.6.4</span> Workflow definieren<a class="anchor" aria-label="anchor" href="#workflow-definieren"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb214"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">knn_fit</span> <span class="op">&lt;-</span> <span class="fu">workflow</span><span class="op">(</span><span class="op">)</span> |&gt;
  <span class="fu">add_recipe</span><span class="op">(</span><span class="va">uc_recipe</span><span class="op">)</span> |&gt;
  <span class="fu">add_model</span><span class="op">(</span><span class="va">knn_spec</span><span class="op">)</span> |&gt;
  <span class="fu">fit</span><span class="op">(</span>data <span class="op">=</span> <span class="va">unscaled_cancer</span><span class="op">)</span>

<span class="va">knn_fit</span></code></pre></div>
<pre><code>## â•â• Workflow [trained] â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•
## Preprocessor: Recipe
## Model: nearest_neighbor()
## 
## â”€â”€ Preprocessor â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
## 2 Recipe Steps
## 
## â€¢ step_scale()
## â€¢ step_center()
## 
## â”€â”€ Model â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
## 
## Call:
## kknn::train.kknn(formula = ..y ~ ., data = data, ks = min_rows(5,     data, 5), kernel = ~"rectangular")
## 
## Type of response variable: nominal
## Minimal misclassification: 0.1107206
## Best kernel: rectangular
## Best k: 5</code></pre>
</div>
<div id="vorhersagen-1" class="section level3" number="6.6.5">
<h3>
<span class="header-section-number">6.6.5</span> Vorhersagen<a class="anchor" aria-label="anchor" href="#vorhersagen-1"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb216"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">new_observation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>Area <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">1500</span><span class="op">)</span>, Smoothness <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.075</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span>
<span class="va">prediction</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">knn_fit</span>, <span class="va">new_observation</span><span class="op">)</span>

<span class="va">prediction</span></code></pre></div>
<pre><code>## # A tibble: 2 Ã— 1
##   .pred_class
##   &lt;fct&gt;      
## 1 B          
## 2 M</code></pre>
</div>
</div>
<div id="mit-train-test-aufteilung" class="section level2" number="6.7">
<h2>
<span class="header-section-number">6.7</span> Mit Train-Test-Aufteilung<a class="anchor" aria-label="anchor" href="#mit-train-test-aufteilung"><i class="fas fa-link"></i></a>
</h2>
<p>Im Kapitel 5 greifen <span class="citation">Timbers, Campbell, and Lee (<a href="references.html#ref-timbers_data_2022" role="doc-biblioref">2022</a>)</span> die Aufteilung in Train- vs.Â Test-Sample noch nicht auf (aber in Kapitel 6).</p>
<p>Da in diesem Kurs diese Aufteilung aber schon besprochen wurde,
soll dies hier auch dargestellt werden.</p>
<div class="sourceCode" id="cb218"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cancer_split</span> <span class="op">&lt;-</span> <span class="fu">initial_split</span><span class="op">(</span><span class="va">cancer</span>, prop <span class="op">=</span> <span class="fl">0.75</span>, strata <span class="op">=</span> <span class="va">Class</span><span class="op">)</span>
<span class="va">cancer_train</span> <span class="op">&lt;-</span> <span class="fu">training</span><span class="op">(</span><span class="va">cancer_split</span><span class="op">)</span>
<span class="va">cancer_test</span> <span class="op">&lt;-</span> <span class="fu">testing</span><span class="op">(</span><span class="va">cancer_split</span><span class="op">)</span> </code></pre></div>
<div id="rezept-definieren-1" class="section level3" number="6.7.1">
<h3>
<span class="header-section-number">6.7.1</span> Rezept definieren<a class="anchor" aria-label="anchor" href="#rezept-definieren-1"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb219"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cancer_recipe</span> <span class="op">&lt;-</span> <span class="fu">recipe</span><span class="op">(</span><span class="va">Class</span> <span class="op">~</span> <span class="va">Smoothness</span> <span class="op">+</span> <span class="va">Concavity</span>, data <span class="op">=</span> <span class="va">cancer_train</span><span class="op">)</span> |&gt;
  <span class="fu">step_scale</span><span class="op">(</span><span class="fu">all_predictors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> |&gt;
  <span class="fu">step_center</span><span class="op">(</span><span class="fu">all_predictors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="modell-definieren-1" class="section level3" number="6.7.2">
<h3>
<span class="header-section-number">6.7.2</span> Modell definieren<a class="anchor" aria-label="anchor" href="#modell-definieren-1"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb220"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">knn_spec</span> <span class="op">&lt;-</span> <span class="fu">nearest_neighbor</span><span class="op">(</span>weight_func <span class="op">=</span> <span class="st">"rectangular"</span>, neighbors <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> |&gt;
  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"kknn"</span><span class="op">)</span> |&gt;
  <span class="fu">set_mode</span><span class="op">(</span><span class="st">"classification"</span><span class="op">)</span></code></pre></div>
</div>
<div id="workflow-definieren-1" class="section level3" number="6.7.3">
<h3>
<span class="header-section-number">6.7.3</span> Workflow definieren<a class="anchor" aria-label="anchor" href="#workflow-definieren-1"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb221"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">knn_fit</span> <span class="op">&lt;-</span> <span class="fu">workflow</span><span class="op">(</span><span class="op">)</span> |&gt;
  <span class="fu">add_recipe</span><span class="op">(</span><span class="va">cancer_recipe</span><span class="op">)</span> |&gt;
  <span class="fu">add_model</span><span class="op">(</span><span class="va">knn_spec</span><span class="op">)</span> |&gt;
  <span class="fu">fit</span><span class="op">(</span>data <span class="op">=</span> <span class="va">cancer_train</span><span class="op">)</span>

<span class="va">knn_fit</span></code></pre></div>
<pre><code>## â•â• Workflow [trained] â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•
## Preprocessor: Recipe
## Model: nearest_neighbor()
## 
## â”€â”€ Preprocessor â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
## 2 Recipe Steps
## 
## â€¢ step_scale()
## â€¢ step_center()
## 
## â”€â”€ Model â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
## 
## Call:
## kknn::train.kknn(formula = ..y ~ ., data = data, ks = min_rows(3,     data, 5), kernel = ~"rectangular")
## 
## Type of response variable: nominal
## Minimal misclassification: 0.09624413
## Best kernel: rectangular
## Best k: 3</code></pre>
</div>
<div id="vorhersagen-2" class="section level3" number="6.7.4">
<h3>
<span class="header-section-number">6.7.4</span> Vorhersagen<a class="anchor" aria-label="anchor" href="#vorhersagen-2"><i class="fas fa-link"></i></a>
</h3>
<p>Im Gegensatz zu <span class="citation">Timbers, Campbell, and Lee (<a href="references.html#ref-timbers_data_2022" role="doc-biblioref">2022</a>)</span> verwenden wir hier <code>last_fit()</code> und <code>collect_metrics()</code>,
da wir dies bereits eingefÃ¼hrt haben und kÃ¼nftig darauf aufbauen werden.</p>
<div class="sourceCode" id="cb223"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cancer_test_fit</span> <span class="op">&lt;-</span> <span class="fu">last_fit</span><span class="op">(</span><span class="va">knn_fit</span>, <span class="va">cancer_split</span><span class="op">)</span>

<span class="va">cancer_test_fit</span></code></pre></div>
<pre><code>## # Resampling results
## # Manual resampling 
## # A tibble: 1 Ã— 6
##   splits            id               .metrics .notes   .predictions .workflow 
##   &lt;list&gt;            &lt;chr&gt;            &lt;list&gt;   &lt;list&gt;   &lt;list&gt;       &lt;list&gt;    
## 1 &lt;split [426/143]&gt; train/test split &lt;tibble&gt; &lt;tibble&gt; &lt;tibble&gt;     &lt;workflow&gt;</code></pre>
</div>
<div id="modellgÃ¼te-1" class="section level3" number="6.7.5">
<h3>
<span class="header-section-number">6.7.5</span> ModellgÃ¼te<a class="anchor" aria-label="anchor" href="#modellg%C3%BCte-1"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb225"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cancer_test_fit</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">collect_metrics</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 2 Ã— 4
##   .metric  .estimator .estimate .config             
##   &lt;chr&gt;    &lt;chr&gt;          &lt;dbl&gt; &lt;chr&gt;               
## 1 accuracy binary         0.853 Preprocessor1_Model1
## 2 roc_auc  binary         0.847 Preprocessor1_Model1</code></pre>
<p>Die eigentlichen Predictions stecken in der Listenspalte <code>.predictions</code> im Fit-Objekt:</p>
<div class="sourceCode" id="cb227"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cancer_test_fit</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "splits"       "id"           ".metrics"     ".notes"       ".predictions"
## [6] ".workflow"</code></pre>
<p>Genau genommen ist <code>.predictions</code> eine Spalte, in der in jeder Zeile (und damit Zelle) eine Tabelle (Tibble) steht.
Wir haben nur eine Zeile und wollen das erste Element dieser Spalte herausziehen.
Da hilft <code><a href="https://purrr.tidyverse.org/reference/pluck.html">pluck()</a></code>:</p>
<div class="sourceCode" id="cb229"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cancer_test_predictions</span> <span class="op">&lt;-</span> 
<span class="va">cancer_test_fit</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html">pluck</a></span><span class="op">(</span><span class="st">".predictions"</span>, <span class="fl">1</span><span class="op">)</span>

<span class="va">confusion</span> <span class="op">&lt;-</span> <span class="va">cancer_test_predictions</span> |&gt;
             <span class="fu">conf_mat</span><span class="op">(</span>truth <span class="op">=</span> <span class="va">Class</span>, estimate <span class="op">=</span> <span class="va">.pred_class</span><span class="op">)</span>

<span class="va">confusion</span></code></pre></div>
<pre><code>##           Truth
## Prediction  B  M
##          B 78  9
##          M 12 44</code></pre>
</div>
<div id="visualisierung" class="section level3" number="6.7.6">
<h3>
<span class="header-section-number">6.7.6</span> Visualisierung<a class="anchor" aria-label="anchor" href="#visualisierung"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb231"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html">autoplot</a></span><span class="op">(</span><span class="va">confusion</span>, type <span class="op">=</span> <span class="st">"mosaic"</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html">autoplot</a></span><span class="op">(</span><span class="va">confusion</span>, type <span class="op">=</span> <span class="st">"heatmap"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Beobachtung"</span>,
       y <span class="op">=</span> <span class="st">"Vorhersage"</span>,
       title <span class="op">=</span> <span class="st">"Konfusionsmatrix"</span><span class="op">)</span></code></pre></div>
<p><img src="chunk-img/tidymodelsunnamed-chunk-18-1.png" width="70%" style="display: block; margin: auto;"><img src="chunk-img/tidymodelsunnamed-chunk-18-2.png" width="70%" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="kennzahlen-der-klassifikation" class="section level2" number="6.8">
<h2>
<span class="header-section-number">6.8</span> Kennzahlen der Klassifikation<a class="anchor" aria-label="anchor" href="#kennzahlen-der-klassifikation"><i class="fas fa-link"></i></a>
</h2>
<p>In <span class="citation">Sauer (<a href="references.html#ref-modar" role="doc-biblioref">2019</a>)</span>, Kap. 19.6, findet sich einige ErklÃ¤rung zu Kennzahlen der KlassifikationsgÃ¼te.</p>
<p>Ein Test kann vier verschiedenen Ergebnisse haben:</p>
<div id="tlgwujqsni" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#tlgwujqsni .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#tlgwujqsni .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#tlgwujqsni .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#tlgwujqsni .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 6px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#tlgwujqsni .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#tlgwujqsni .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#tlgwujqsni .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#tlgwujqsni .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#tlgwujqsni .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#tlgwujqsni .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#tlgwujqsni .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#tlgwujqsni .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#tlgwujqsni .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#tlgwujqsni .gt_from_md > :first-child {
  margin-top: 0;
}

#tlgwujqsni .gt_from_md > :last-child {
  margin-bottom: 0;
}

#tlgwujqsni .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#tlgwujqsni .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#tlgwujqsni .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#tlgwujqsni .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#tlgwujqsni .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#tlgwujqsni .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#tlgwujqsni .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#tlgwujqsni .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#tlgwujqsni .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#tlgwujqsni .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#tlgwujqsni .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#tlgwujqsni .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#tlgwujqsni .gt_left {
  text-align: left;
}

#tlgwujqsni .gt_center {
  text-align: center;
}

#tlgwujqsni .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#tlgwujqsni .gt_font_normal {
  font-weight: normal;
}

#tlgwujqsni .gt_font_bold {
  font-weight: bold;
}

#tlgwujqsni .gt_font_italic {
  font-style: italic;
}

#tlgwujqsni .gt_super {
  font-size: 65%;
}

#tlgwujqsni .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 65%;
}
</style>
<div class="inline-table"><table class="gt_table">
<caption>
<span id="tab:class-stats">Table 6.1: </span>Vier Arten von Ergebnissen von Klassifikationen</caption>
  
  <thead class="gt_col_headings"><tr>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Wahrheit</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Als negativ (-) vorhergesagt</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Als positiv (+) vorhergesagt</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Summe</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td class="gt_row gt_left">In Wahrheit negativ (-)</td>
<td class="gt_row gt_left">Richtig negativ (RN)</td>
<td class="gt_row gt_left">Falsch positiv (FP)</td>
<td class="gt_row gt_left">N</td>
</tr>
<tr>
<td class="gt_row gt_left">In Wahrheit positiv (+)</td>
<td class="gt_row gt_left">Falsch negativ (FN)</td>
<td class="gt_row gt_left">Richtig positiv (RN)</td>
<td class="gt_row gt_left">P</td>
</tr>
<tr>
<td class="gt_row gt_left">Summe</td>
<td class="gt_row gt_left">N*</td>
<td class="gt_row gt_left">P*</td>
<td class="gt_row gt_left">N+P</td>
</tr>
</tbody>
</table></div>
</div>
<p>Es gibt eine verwirrende Vielfalt von Kennzahlen,
um die GÃ¼te einer Klassifikation einzuschÃ¤tzen.
Hier sind einige davon:</p>
<div id="hgahrvcueg" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#hgahrvcueg .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#hgahrvcueg .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#hgahrvcueg .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#hgahrvcueg .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 6px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#hgahrvcueg .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#hgahrvcueg .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#hgahrvcueg .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#hgahrvcueg .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#hgahrvcueg .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#hgahrvcueg .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#hgahrvcueg .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#hgahrvcueg .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#hgahrvcueg .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#hgahrvcueg .gt_from_md > :first-child {
  margin-top: 0;
}

#hgahrvcueg .gt_from_md > :last-child {
  margin-bottom: 0;
}

#hgahrvcueg .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#hgahrvcueg .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#hgahrvcueg .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#hgahrvcueg .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#hgahrvcueg .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#hgahrvcueg .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#hgahrvcueg .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#hgahrvcueg .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#hgahrvcueg .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#hgahrvcueg .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#hgahrvcueg .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#hgahrvcueg .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#hgahrvcueg .gt_left {
  text-align: left;
}

#hgahrvcueg .gt_center {
  text-align: center;
}

#hgahrvcueg .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#hgahrvcueg .gt_font_normal {
  font-weight: normal;
}

#hgahrvcueg .gt_font_bold {
  font-weight: bold;
}

#hgahrvcueg .gt_font_italic {
  font-style: italic;
}

#hgahrvcueg .gt_super {
  font-size: 65%;
}

#hgahrvcueg .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 65%;
}
</style>
<div class="inline-table"><table class="gt_table">
<thead class="gt_header">
<tr>
<th colspan="3" class="gt_heading gt_title gt_font_normal" style>GelÃ¤ufige Kennwerte der Klassifikation.</th>
    </tr>
<tr>
<th colspan="3" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border" style>F: Falsch. R: Richtig. P: Positiv. N: Negativ</th>
    </tr>
</thead>
<thead class="gt_col_headings"><tr>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Name</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Definition</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Synonyme</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td class="gt_row gt_left">FP-Rate</td>
<td class="gt_row gt_left">FP/N</td>
<td class="gt_row gt_left">Alphafehler, Typ-1-Fehler, 1-SpezifitÃ¤t, Fehlalarm</td>
</tr>
<tr>
<td class="gt_row gt_left">RP-Rate</td>
<td class="gt_row gt_left">RP/N</td>
<td class="gt_row gt_left">Power, SensitivitÃ¤t, 1-Betafehler, Recall</td>
</tr>
<tr>
<td class="gt_row gt_left">FN-Rate</td>
<td class="gt_row gt_left">FN/N</td>
<td class="gt_row gt_left">Fehlender Alarm, Befafehler</td>
</tr>
<tr>
<td class="gt_row gt_left">RN-Rate</td>
<td class="gt_row gt_left">RN/N</td>
<td class="gt_row gt_left">SpezifitÃ¤t, 1-Alphafehler</td>
</tr>
<tr>
<td class="gt_row gt_left">Pos. Vorhersagewert</td>
<td class="gt_row gt_left">RP/P*</td>
<td class="gt_row gt_left">PrÃ¤zision, Relevanz</td>
</tr>
<tr>
<td class="gt_row gt_left">Neg. Vorhersagewert</td>
<td class="gt_row gt_left">RN/N*</td>
<td class="gt_row gt_left">Segreganz</td>
</tr>
<tr>
<td class="gt_row gt_left">Richtigkeit</td>
<td class="gt_row gt_left">(RP+RN)/(N+P)</td>
<td class="gt_row gt_left">Korrektklassifikationsrate, Gesamtgenauigkeit</td>
</tr>
</tbody>
</table></div>
</div>
</div>
<div id="krebstest-beispiel" class="section level2" number="6.9">
<h2>
<span class="header-section-number">6.9</span> Krebstest-Beispiel<a class="anchor" aria-label="anchor" href="#krebstest-beispiel"><i class="fas fa-link"></i></a>
</h2>
<p>Betrachten wir Daten eines fiktiven Krebstest, aber realistischen
Daten.</p>
<pre><code>## # A tibble: 1 Ã— 7
##   format width height colorspace matte filesize density
##   &lt;chr&gt;  &lt;int&gt;  &lt;int&gt; &lt;chr&gt;      &lt;lgl&gt;    &lt;int&gt; &lt;chr&gt;  
## 1 PNG      500    429 sRGB       TRUE     40643 72x72</code></pre>
<div class="inline-figure"><img src="chunk-img/tidymodelsunnamed-chunk-19-1.png" width="70%" style="display: block; margin: auto;"></div>
<p>Wie gut ist dieser Test?
Berechnen wir einige Kennzahlen.</p>
<p>Da die Funktionen zur Klassifikation stets einen Faktor wollen,
wandeln wir die relevanten Spalten zuerst in einen Faktor um (aktuell sind es numerische Spalten).</p>
<div class="sourceCode" id="cb233"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">krebstest</span> <span class="op">&lt;-</span>
  <span class="va">krebstest</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Krebs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">Krebs</span><span class="op">)</span>,
         Test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">Test</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Gesamtgenauigkeit:</p>
<div class="sourceCode" id="cb234"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">accuracy</span><span class="op">(</span><span class="va">krebstest</span>, truth <span class="op">=</span> <span class="va">Krebs</span>, estimate <span class="op">=</span> <span class="va">Test</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 1 Ã— 3
##   .metric  .estimator .estimate
##   &lt;chr&gt;    &lt;chr&gt;          &lt;dbl&gt;
## 1 accuracy binary          0.87</code></pre>
<p>SensitivitÃ¤t:</p>
<div class="sourceCode" id="cb236"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">sens</span><span class="op">(</span><span class="va">krebstest</span>, truth <span class="op">=</span> <span class="va">Krebs</span>, estimate <span class="op">=</span> <span class="va">Test</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 1 Ã— 3
##   .metric .estimator .estimate
##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;
## 1 sens    binary         0.884</code></pre>
<p>SpezifitÃ¤t:</p>
<div class="sourceCode" id="cb238"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://readr.tidyverse.org/reference/spec.html">spec</a></span><span class="op">(</span><span class="va">krebstest</span>, truth <span class="op">=</span> <span class="va">Krebs</span>, estimate <span class="op">=</span> <span class="va">Test</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 1 Ã— 3
##   .metric .estimator .estimate
##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;
## 1 spec    binary           0.6</code></pre>
<p>Kappa:</p>
<div class="sourceCode" id="cb240"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">kap</span><span class="op">(</span><span class="va">krebstest</span>, truth <span class="op">=</span> <span class="va">Krebs</span>, estimate <span class="op">=</span> <span class="va">Test</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 1 Ã— 3
##   .metric .estimator .estimate
##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;
## 1 kap     binary         0.261</code></pre>
<p>Positiver Vorhersagewert:</p>
<div class="sourceCode" id="cb242"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ppv</span><span class="op">(</span><span class="va">krebstest</span>, truth <span class="op">=</span> <span class="va">Krebs</span>, estimate <span class="op">=</span> <span class="va">Test</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 1 Ã— 3
##   .metric .estimator .estimate
##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;
## 1 ppv     binary         0.977</code></pre>
<p>Negativer Vorhersagewert:</p>
<div class="sourceCode" id="cb244"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">npv</span><span class="op">(</span><span class="va">krebstest</span>, truth <span class="op">=</span> <span class="va">Krebs</span>, estimate <span class="op">=</span> <span class="va">Test</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 1 Ã— 3
##   .metric .estimator .estimate
##   &lt;chr&gt;   &lt;chr&gt;          &lt;dbl&gt;
## 1 npv     binary         0.214</code></pre>
<p>WÃ¤hrend SensitivitÃ¤t und SpezitivitÃ¤t sehr hoch sind,
ist die der negative Vorhersagewert sehr gering:</p>
<p>Wenn man einen positiven Test erhÃ¤lt, ist die
Wahrscheinlichkeit, in Wahrheit krank zu sein gering, zum GlÃ¼ck!</p>
<!-- ## Aufgaben und Vertiefung -->
</div>
<div id="aufgaben-3" class="section level2" number="6.10">
<h2>
<span class="header-section-number">6.10</span> Aufgaben<a class="anchor" aria-label="anchor" href="#aufgaben-3"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>Arbeiten Sie sich so gut als mÃ¶glich durch <a href="https://github.com/sebastiansauer/covid-icu">diese Analyse zum Verlauf von Covid-FÃ¤llen</a>
</li>
<li><a href="https://onezero.blog/modelling-binary-logistic-regression-using-tidymodels-library-in-r-part-1/">Fallstudie zur Modellierung einer logististischen Regression mit tidymodels</a></li>
<li><a href="https://juliasilge.com/blog/multinomial-volcano-eruptions/">Fallstudie zu VulkanausbrÃ¼chen</a></li>
<li><a href="https://juliasilge.com/blog/himalayan-climbing/">Fallstudie Himalaya</a></li>
</ul>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="tidymodels.html"><span class="header-section-number">5</span> tidymodels</a></div>
<div class="next"><a href="resampling-und-tuning.html"><span class="header-section-number">7</span> Resampling und Tuning</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#knn"><span class="header-section-number">6</span> kNN</a></li>
<li>
<a class="nav-link" href="#lernsteuerung-3"><span class="header-section-number">6.1</span> Lernsteuerung</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#lernziele-4"><span class="header-section-number">6.1.1</span> Lernziele</a></li>
<li><a class="nav-link" href="#literatur-4"><span class="header-section-number">6.1.2</span> Literatur</a></li>
</ul>
</li>
<li><a class="nav-link" href="#%C3%BCberblick-1"><span class="header-section-number">6.2</span> Ãœberblick</a></li>
<li><a class="nav-link" href="#intuitive-erkl%C3%A4rung"><span class="header-section-number">6.3</span> Intuitive ErklÃ¤rung</a></li>
<li><a class="nav-link" href="#krebsdiagnostik"><span class="header-section-number">6.4</span> Krebsdiagnostik</a></li>
<li><a class="nav-link" href="#berechnung-der-n%C3%A4he"><span class="header-section-number">6.5</span> Berechnung der NÃ¤he</a></li>
<li>
<a class="nav-link" href="#knn-mit-tidymodels"><span class="header-section-number">6.6</span> kNN mit Tidymodels</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#analog-zu-timbers-et-al."><span class="header-section-number">6.6.1</span> Analog zu Timbers et al.</a></li>
<li><a class="nav-link" href="#rezept-definieren"><span class="header-section-number">6.6.2</span> Rezept definieren</a></li>
<li><a class="nav-link" href="#modell-definieren"><span class="header-section-number">6.6.3</span> Modell definieren</a></li>
<li><a class="nav-link" href="#workflow-definieren"><span class="header-section-number">6.6.4</span> Workflow definieren</a></li>
<li><a class="nav-link" href="#vorhersagen-1"><span class="header-section-number">6.6.5</span> Vorhersagen</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#mit-train-test-aufteilung"><span class="header-section-number">6.7</span> Mit Train-Test-Aufteilung</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#rezept-definieren-1"><span class="header-section-number">6.7.1</span> Rezept definieren</a></li>
<li><a class="nav-link" href="#modell-definieren-1"><span class="header-section-number">6.7.2</span> Modell definieren</a></li>
<li><a class="nav-link" href="#workflow-definieren-1"><span class="header-section-number">6.7.3</span> Workflow definieren</a></li>
<li><a class="nav-link" href="#vorhersagen-2"><span class="header-section-number">6.7.4</span> Vorhersagen</a></li>
<li><a class="nav-link" href="#modellg%C3%BCte-1"><span class="header-section-number">6.7.5</span> ModellgÃ¼te</a></li>
<li><a class="nav-link" href="#visualisierung"><span class="header-section-number">6.7.6</span> Visualisierung</a></li>
</ul>
</li>
<li><a class="nav-link" href="#kennzahlen-der-klassifikation"><span class="header-section-number">6.8</span> Kennzahlen der Klassifikation</a></li>
<li><a class="nav-link" href="#krebstest-beispiel"><span class="header-section-number">6.9</span> Krebstest-Beispiel</a></li>
<li><a class="nav-link" href="#aufgaben-3"><span class="header-section-number">6.10</span> Aufgaben</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/sebastiansauer/datascience1/blob/master/070-knn.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/sebastiansauer/datascience1/edit/master/070-knn.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>DataScience1</strong>: Grundlagen der Prognosemodellierung ğŸ”®ğŸ§°" was written by Sebastian Sauer. It was last built on 2022-05-05 16:57:10.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
